Document C# union types in ASP.NET Core (.NET 11) (#37261)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/unions.md

---
title: Use C# union types in ASP.NET Core
author: DeagleGross
description: Learn how C# union types are supported through System.Text.Json in ASP.NET Core Minimal APIs, MVC, SignalR, Blazor, and OpenAPI.
monikerRange: '>= aspnetcore-11.0'
ms.author: wpickett
ms.date: 06/19/2026
uid: fundamentals/unions
ai-usage: ai-assisted
---

> # Use C# union types in ASP.NET Core

Reviewer: ```suggestion

# Use C# union types in ASP.NET Core
```
Needed an empty line between frontmatter and H1

ASP.NET Core support for unions builds entirely on union support in [`System.Text.Json`](/dotnet/standard/serialization/system-text-json/overview) (STJ). As a result, unions are supported only where STJ is the serializer: the **JSON request body** (input) and the **JSON response body** (output). The same STJ foundation also enables unions in SignalR's `JsonHubProtocol` and in Blazor's JavaScript interop, persisted component state, and prerendered component parameters. Unions aren't supported for non-body binding sources, such as query strings, route values, headers, and form fields. For more information, see [Where unions aren't supported](#where-unions-arent-supported).

## How unions are serialized

A union type is declared with the `union` keyword and a list of case types:

```csharp

> public union UnionIntString(int, string);

Reviewer: ```suggestion
public union UnionIntString(int, string);
public union UnionBoolString(bool, string);
public union UnionNullableIntString(int?, string);
```
These additional two were not declared.  Should they be here?

Author: ye, we can add those. Will be easier for demonstration

                reader.Skip(); // Skip the value.
            }

            return null; // No case identified; deserialization fails.
        };
    }
}
```

> [!NOTE]
> `JsonUnionAttribute`, `JsonTypeClassifier`, and `JsonTypeClassifierFactory<T>` are new `System.Text.Json` APIs in .NET 11. A classifier can also be registered for multiple union types through <xref:System.Text.Json.JsonSerializerOptions>.
<!-- TODO: Add <xref:> links for the .NET 11 System.Text.Json union APIs (JsonUnionAttribute, JsonTypeClassifier, JsonTypeClassifierFactory<T>) once published to dotnet-api-docs. -->

## Minimal APIs

Unions work as request body parameters and as return types in both the runtime path (`RequestDelegateFactory`) and the source-generated [Request Delegate Generator (RDG)](xref:fundamentals/aot/rdg). Behavior is identical across both.

```csharp
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

// Request body: UnionBoolString is unambiguous (bool vs string),
// so it binds without a classifier.
app.MapPost("/flag", (UnionBoolString flag) => flag);

// Request body: UnionPet is object-cased, so it needs the classifier

> // shown earlier to bind from a body.

Reviewer: ```suggestion
// Request body: UnionPet's cases are both objects (StartObject token),
// so it needs the classifier shown earlier to bind from a body.
```
I think this might translate better when localized.  Object-cased isn't as common terminology.

### Deserialization uses the first JSON token

Because there's no discriminator on the wire, STJ recovers the case type from the JSON value itself by inspecting the *first token* and selecting the single case whose declared type is compatible with that token kind. The mapping is fixed:

| <xref:System.Text.Json.JsonTokenType> | Compatible case types |
| --- | --- |
| `Number` | Numeric primitives, such as `int`, `long`, `double`, and `decimal` |
| `String` | `string`, `DateTime`, `DateTimeOffset`, `Guid`, `TimeSpan`, `Uri`, `char`, `byte[]`, and enums |
| `True` / `False` | `bool` |
| `StartObject` | Objects and dictionaries |
| `StartArray` | Arrays and collections |
| `Null` | `null` |

Selection is O(1) and requires no read-ahead. A union whose cases map to distinct token kinds, such as `UnionBoolString(bool, string)`, deserializes without any extra configuration.

> [!IMPORTANT]

> > Minimal APIs and MVC serialize and deserialize HTTP JSON using web defaults (<xref:System.Text.Json.JsonSerializerDefaults>.Web), which enable `JsonNumberHandling.AllowReadingFromString`. This lets a numeric case be read from a JSON `String` token, so the `String` token becomes compatible with *both* numeric cases and `string`. As a result, a union that pairs a numeric case with a `string` case—such as `UnionIntString(int, string)`—is ambiguous on the `String` token for HTTP JSON binding. A `Number` payload binds to the numeric case, but a `String` payload returns `400 Bad Request` unless the union supplies a classifier. This differs from plain `System.Text.Json` and from SignalR's `JsonHubProtocol`, which deserialize the same union without a classifier.

Reviewer: ```suggestion
> Minimal APIs and MVC serialize and deserialize HTTP JSON using <xref:System.Text.Json.JsonSerializerDefaults.Web?displayProperty=nameWithType>, which enables `JsonNumberHandling.AllowReadingFromString`. This lets a numeric case be read from a JSON `String` token, so the `String` token becomes compatible with *both* numeric cases and `string`. As a result, a union that pairs a numeric case with a `string` case—such as `UnionIntString(int, string)`—is ambiguous on the `String` token for HTTP JSON binding. A `Number` payload binds to the numeric case, but a `String` payload returns `400 Bad Request` unless the union supplies a classifier. This differs from plain `System.Text.Json` and from SignalR's `JsonHubProtocol`, which deserialize the same union without a classifier.
```
Minor: Link fix and verb change to "enables"

    {
        "value" => new UnionIntString(42),
        "string" => new UnionIntString("hi"),
        _ => default,
    };
}
```

The same serialization rules apply: only the active case is written, and ambiguous unions need a classifier to deserialize.

## SignalR

`JsonHubProtocol` forwards reads and writes to `System.Text.Json`, so unions work as hub method parameters, return values, and stream items without any extra configuration:

```csharp
public class ChatHub : Hub
{
    // Union argument (client → server).
    public Task Send(UnionIntString message) => Clients.All.SendAsync("Receive", message);

    // Union return value (server → client).
    public UnionPet GetPet() => new UnionPet(new Cat("Whiskers"));

    // Union stream items (server → client).
    public async IAsyncEnumerable<UnionIntString> Stream()
    {
        yield return 1;
        yield return "two";
    }
}
```

On the read path, the parameter, return, or stream-item `Type` resolved from the invocation binder drives the union converter, including any `[JsonUnion]` classifier. On the write path, the active case is serialized with no envelope or discriminator.

> Unlike HTTP JSON binding in Minimal APIs and MVC, `JsonHubProtocol` doesn't treat a JSON `String` token as ambiguous for numeric cases, so a union such as `UnionIntString(int, string)` round-trips both the `int` and `string` cases without a classifier. Object-cased unions, such as `UnionPet(Cat, Dog)`, are still ambiguous on read and require a classifier.

Reviewer: ```suggestion
Unlike HTTP JSON binding in Minimal APIs and MVC, `JsonHubProtocol` doesn't treat a JSON `String` token as ambiguous for numeric cases, so a union such as `UnionIntString(int, string)` round-trips both the `int` and `string` cases without a classifier. Unions whose cases share the `StartObject` token, such as `UnionPet(Cat, Dog)`, are still ambiguous on read and require a classifier.
```
Very minor: Better I think to use the same vocabulary the article introduces earlier ("share a JSON token kind") instead of new jargon "object-cased".

## OpenAPI

The OpenAPI document represents a union as an [`anyOf`](https://spec.openapis.org/oas/latest#composition-and-inheritance-polymorphism) schema, with one entry per case type:

```json
"Cat": {
  "type": "object",
  "properties": {
    "name": { "type": "string" }
  }
},
"Dog": {
  "type": "object",
  "properties": {
    "breed": { "type": "string" }
  }
},
"UnionIntString": {
  "anyOf": [
    { "type": "integer", "format": "int32" },
    { "type": "string" }
  ]
},
"UnionPet": {
  "type": "object",
  "anyOf": [
    { "$ref": "#/components/schemas/Cat" },
    { "$ref": "#/components/schemas/Dog" }
  ]
}
```

Because a union case has no discriminator and is structurally identical to the standalone type, each case schema reuses the standalone component name. The `Cat` and `Dog` schemas referenced by `UnionPet` are the same components that a standalone `Cat` or `Dog` endpoint produces. This differs from polymorphic types, whose derived schemas are lifted to prefixed component names because they carry a `$type` discriminator.

> An endpoint can also produce multiple response types for the same status code and content type. <xref:Microsoft.AspNetCore.Mvc.ApiExplorer> preserves every declared response type, and the generated document emits an `anyOf` schema when several types share a content type:

Reviewer: ```suggestion
An endpoint can also produce multiple response types for the same status code and content type. The <xref:Microsoft.AspNetCore.Mvc.ApiExplorer?displayProperty=fullName> namespace preserves every declared response type, and the generated document emits an `anyOf` schema when several types share a content type:
```
Very minor: Added `fullName` so it did not end up looking like just "ApiExplorer" which might be assumed to be a type rather than a namespace.

```csharp
app.MapGet("/any-of", () => Results.Ok())
    .Produces<UnionPet>(StatusCodes.Status200OK, "application/json")

>     .Produces<Clinic>(StatusCodes.Status200OK, "application/json");

Reviewer: ```suggestion
    .Produces<Cat>(StatusCodes.Status200OK, "application/json");
```
Very minor:  "Clinic" isn't mentioned before this use and wasn't declared.  Cat was.  We couild use one already declared.

Author: `Cat` is a case-type for `UnionPet` so it is not the ideal example. I have added a new union here with definition to make it more "meaningful"

```suggestion
    .Produces<UnionDinosaur>(StatusCodes.Status200OK, "application/json");
    
public record Tyrannosaurus(string Name, double BiteForceNewtons);
public record Triceratops(string Name, int HornCount);
public record Velociraptor(string Name, double TopSpeedKmh);
public union UnionDinosaur(Tyrannosaurus, Triceratops, Velociraptor);
```
